Avoid accidental deletions of lock files on write errors

alalazo opened the pull request:
fixes #18441

Modifications:

- [x] Added a context manager to ensure that files or dirs are safely removed only if the block in the context executes successfully
- [x] Employ that context manager in `Environment.write` to avoid accidental deletions of `spack.lock`
- [x] Add unit tests to stress the context manager and avoid regressions

alalazo committed:
Lockfile deletion: added test case to reproduce failure

alalazo committed:
Fix lockfile deletion issue

fixes #18441

A new context manager that ensures safe deletion
of files or dirs has been added. This context manager
is used whenever writing Spack environments, to avoid
accidental deletion of lock files in case of errors.

alalazo commented:
This part:

https://github.com/spack/spack/blob/73a2ac12b56eb9c31baeed59e68f5b9a15bb2a48/lib/spack/spack/environment.py#L1496-L1520

can maybe be refactored better, but it's not fully clear to me when (i.e. under which circumstances or use cases) one branch of the conditional is taken with respect to the other. So, I currently opted for a refactoring that is not altering the flow of the code.

scheibelp commented on lib/spack/spack/environment.py:
```diff
@@ -1514,13 +1514,30 @@ def write(self, regenerate_views=True):
+        checks have been successful.
+
+        Args:
+            raw_yaml_dict: raw YAML dictionary
```
These parameter descriptions don't add anything - they are not needed.

alalazo replied:
Done

scheibelp commented on lib/spack/spack/environment.py:
```diff
@@ -1514,13 +1514,30 @@ def write(self, regenerate_views=True):
+        if regenerate_views:
+            self.regenerate_views()
+
+    def _write_helper(self, raw_yaml_dict, yaml_dict):
```
I think a more specific name would be helpful: `_update_and_write_manifest` perhaps?

The docstring could elaborate:

```
Update YAML manifest for this environment based on changes to spec lists and views and write it.
```

alalazo replied:
Done

scheibelp commented on lib/spack/spack/test/llnl/util/filesystem.py:
```diff
@@ -484,3 +484,57 @@ def test_filter_files_multiple(tmpdir):
+    [{'subdirs': ['myenv'],
+      'files': ['spack.lock', 'spack.yaml'],
+     'remove': ['myenv']}],
+    # Combine a mix if directories and files
```
(typo): mix *of* directories and files

scheibelp commented on lib/spack/spack/test/llnl/util/filesystem.py:
```diff
@@ -484,3 +484,57 @@ def test_filter_files_multiple(tmpdir):
+        # Create files in the directory
+        files = entry['files']
+        for f in files:
+            abspath = str(dir.join(f))
```
What happens if there are multiple directories in `subdirs`? The parameterization appears to be quite flexible, but I'm not sure if the test handles that (or needs to handle it).

`test_activations.python_and_extension_dirs` uses another approach for specifying file/directory hierarchies that may be useful.

alalazo replied:
The idea is to create subdirectories. I'll add tests for that to ensure the context manager is working properly.

alalazo replied:
Added unit tests to mock nested subdirectories.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -974,6 +974,43 @@ def remove_linked_tree(path):
+    them in case any exception is raised in the context block.
+
+    Filenames clash from single files in different directories are currently
+    not handled.
```
Can you give an example of this?

Also, I think this function could be simplified by requiring absolute paths rather than handling `glob`ing for the user.

alalazo replied:
I'll lift this limitation while improving the function according to review.

alalazo replied:
Removed these two lines since the limitations is not there anymore. This is the test for the case above:

https://github.com/spack/spack/blob/9063f3b5b80a5da9afde082a2497663d64a2bae5/lib/spack/spack/test/llnl/util/filesystem.py#L523-L529

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -974,6 +974,43 @@ def remove_linked_tree(path):
+            shutil.move(src, dst_root)
+            removed[src] = dst
+        yield removed
+    except Exception:
```
`Exception` excludes `KeyboardInterrupt` - are there some exceptions that should bypass this guard? I would have expected `except:`.

alalazo replied:
I'll go with `except BaseException:` to appease flake8

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -974,6 +974,43 @@ def remove_linked_tree(path):
+    removed, dst_root = {}, tempfile.mkdtemp()
+
+    try:
+        for file_or_dir in itertools.chain(*glob_matches):
```
It occurs to me that if you have a mixture of files or directories where some file is contained in one of the directories to remove, there could be an issue:

`[x/, x/y]`

it looks like in this case `x/` would be removed before `x/y` and there would be an error when you got to `x/y`.

I think requiring that the user provide absolute paths would simplify this issue (though you may have to sort and check).

alalazo replied:
I don't think we need to require absolute path, but this is definitely a case worth testing that is failing in the current implementation. Thanks for spotting it!

alalazo replied:
I improved the context manager to handle this case and added this unit test:

https://github.com/spack/spack/blob/9063f3b5b80a5da9afde082a2497663d64a2bae5/lib/spack/spack/test/llnl/util/filesystem.py#L530-L533

scheibelp reviewed (changes requested):
This generally looks good but I have some questions/requests.

alalazo committed:
Improved the "safe_remove" context manager

* Test the case of nested subdirectories
* Test the case of removing first a directory,
  then a file in that directory

alalazo commented:
@scheibelp Ready for another review!

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -974,6 +974,49 @@ def remove_linked_tree(path):
+        yield removed
+    except BaseException:
+        # Restore the files that were removed
+        for src, dst in removed.items():
```
(minor) Could you use a name like `orig_path` instead of `src`, and `tmp_path` instead of `dst`? I think that would be slightly more clear but feel free to skip that if you disagree.

alalazo replied:
Done in 8982ee4

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -974,6 +974,49 @@ def remove_linked_tree(path):
+            # This is the case, for instance, if we remove the directory
+            # "/foo/bar" before the file "/foo/bar/baz.yaml".
+            if not os.path.exists(src):
+                continue
```
This handles the first part where you have a removal order like `[x/, x/y]`. However, if you have a removal order like `[x/y, x/]`, then in the `except:` logic, if you try to add `x/y` back before `x/` you could encounter an error.

Sorting (considering shorter paths first) would resolve this (but you'd have to take the abspath of everything first).

alalazo replied:
Done in 8982ee4

scheibelp commented on lib/spack/spack/test/llnl/util/filesystem.py:
```diff
@@ -484,3 +484,82 @@ def test_filter_files_multiple(tmpdir):
+
+
+# Each test input is a tuple of entries which prescribe
+# - the 'subdirs' to be created from tmpdir
```
I would have expected this to just be specified like "x/y" but I'm guessing you are doing this because `tmpdir.ensure` wouldn't work if you passed it `x/y`. To make this clearer, could you have a single argument (i.e. use "x/y" as a parameter instead of ["x", "y"]) but then do something like

```
path.split(os.pathsep)
```

which would decompose the path into the individual directory args.

alalazo replied:
> but I'm guessing you are doing this because tmpdir.ensure wouldn't work if you passed it x/y

That's correct. I'm writing tests parameters like that because they're ready to be consumed by `tmpdir.ensure`. I'll change like you said (though I'll split on `/` and not on `os.sep` so that I don't have to construct the parameter with `os.path.join`, which would defeat the purpose of added clarity).

alalazo replied:
Done in 8982ee4

scheibelp reviewed (changes requested):
I have a few more requests.

alalazo committed:
safe_remove: order paths by their length

This ensures we can treat the case where we request
deletion of a file in a directory and then of the
directory itself.

alalazo commented:
@scheibelp I think requests have been taken care of, ready for another review iteration.

scheibelp commented on lib/spack/spack/test/llnl/util/filesystem.py:
```diff
@@ -484,3 +484,88 @@ def test_filter_files_multiple(tmpdir):
+     {'subdirs': 'work/myenv2',
+      'files': ['spack.lock', 'spack.yaml'],
+      'remove': ['work/myenv2']}],
+    # Remove files with the same name in different directories
```
My final remaining concern is that I'm not sure this actually tests whether the files are restored to the proper location (e.g. if the context manager mixes up the files) - from a testing perspective I think the best way to ensure that is to put different contents in each of the files.

alalazo replied:
I'll write an ad-hoc test for that.

alalazo replied:
Done in 035ab78

alalazo committed:
safe_remove: added an additional test

Added a test to ensure that the context manager is
behaving correctly for files with the same name, by
checking their content.

alalazo commented:
@scheibelp Added an ad-hoc test to check that files with the same name are restored correctly.

scheibelp commented:
Thanks!

alalazo commented:
Thanks for the review!
